Detectors/FOCAL/simulation/src/Detector.cxx

```diff
@@ -1117,9 +1119,49 @@ bool Detector::ProcessHitsHCAL(FairVolume* v)
 
 bool Detector::ProcessHitsEPix(FairVolume* v)
 {
-  LOG(debug) << "We are in sensitive volume " << v->GetName() << ": " << TVirtualMC::GetMC()->CurrentVolPath();
-  //
-  // TODO: to be filled with body
-  //
+  LOG(debug) << "We are in sensitive volume " << v->GetName() << ": " << fMC->CurrentVolPath();
+
+  double eloss = fMC->Edep() * 1e9; // energy in eV  (GeV->eV)
+  if (eloss < DBL_EPSILON) {
+    return false; // only process hits which actually deposit some energy in the FOCAL
+  }
+
+  // In case of new parent track create new track reference
+  auto o2stack = static_cast<o2::data::Stack*>(fMC->GetStack());
+  if (!mCurrentSuperparent->mHasTrackReference) {
+    float x, y, z, px, py, pz, e;
+    fMC->TrackPosition(x, y, z);
+    fMC->TrackMomentum(px, py, pz, e);
+    o2::TrackReference trackref(x, y, z, px, py, pz, fMC->TrackLength(), fMC->TrackTime(), mCurrentParentID, GetDetId());
+    o2stack->addTrackReference(trackref);
+    mCurrentSuperparent->mHasTrackReference = true;
+  }
+
+  float posX, posY, posZ;
+  fMC->TrackPosition(posX, posY, posZ);
+
+  auto [indetector, col, row, layer, segment] = mGeometry->getVirtualInfo(posX, posY, posZ);
+
+  if (!indetector) {
+    // particle outside the detector
+    return true;
+  }
+
+  auto currenthit = FindHit(mCurrentParentID, col, row, layer);
+  if (!currenthit) {
+    // Condition for new hit:
+    // - Processing different partent track (parent track must be produced outside FOCAL)
+    // - Inside different cell
+    // - First track of the event
+    Double_t time = fMC->TrackTime() * 1e9; // time in ns
+    LOG(debug3) << "Adding new hit for parent " << mCurrentParentID << " and cell Col: " << col << " Row: " << row << " segment: " << segment;
+
+    /// check handling of primary particles
+    AddHit(mCurrentParentID, mCurrentPrimaryID, mCurrentSuperparent->mEnergy, row * col + col, o2::focal::Hit::Subsystem_t::EPIXELS, math_utils::Point3D<float>(posX, posY, posZ), time, eloss);
+    o2stack->addHit(GetDetId());
+  } else {
+    LOG(debug3) << "Adding energy to the current hit";
+    currenthit->SetEnergyLoss(currenthit->GetEnergyLoss() + eloss);
+  }
```

Comment:
The idea to sum the energy is particularly appropriate for large areas, like the pads or HCAL towers. For the pixel I don't think it will work, because like this you sum up the energy of the whole chip in one hit (same tower index).
The minimal change is that you do the FindHit per pixel index (which will likely always be false, due to the segmentation). We have here two options:
a) distribute energy to pixels in Detector simulation, and have hits per pixel
b) distribute energy to pixels in Digitizer and have hits per chip with local coordinate
The second option is how I see it implemented. Not sure the first one wouldn't be the more optimal one. However in case we distribute the energy to pixels here the code will become a lot more difficult.

Reply:
As mentioned above, here we have areas of 500 microns, which is a "large" area. So we will have multiple hits per each area, in particular in the shower core. If I remember correctly, for high energy showers, we will reach saturation even with the realistic small pixels, which means we will have relatively large fractions of pixels with more than one hit per pixel. The difussion will also increase occupancy, so this will then happen even more.
But I agree with you regarding the distribution of the energy to pixels (difussion) to be done in the digitizer, so option b.